Support custom styles for resource style

davidfowl opened the pull request:
- Added support for sending custom style to the dashboard via the resource server protocol.
- Use this for azure resource to show more descriptive state updates.

Contributes to https://github.com/dotnet/aspire/issues/2590
Contributes to https://github.com/dotnet/aspire/issues/2415
Fixes https://github.com/dotnet/aspire/issues/2816

davidfowl committed:
Support custom styles for resource style
- Added support for sending custom style to the dashboard via the resource server protocol.
- Use this for azure resource to show more descriptive state updates.

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -155,6 +155,11 @@ message Resource {
     repeated Url urls = 13;
+
+    // The style of the state. This is used to determine the state icon.
+    // Supported styles are success", "warning", "info". Any other style
```
```suggestion
    // Supported styles are "success", "warning", "info". Any other style
```

drewnoakes commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -155,6 +155,11 @@ message Resource {
     // The list of urls that this resource exposes
     repeated Url urls = 13;
+
+    // The style of the state. This is used to determine the state icon.
```
Did you consider encoding a FluentUI icon name (e.g. `ErrorCircle`), style (e.g. `Filled`) and colour (e.g. `Blue`) here instead? That would give a lot more control over the UI to the server.

davidfowl replied:
I would prefer that actually. So just parse the enum?

drewnoakes replied:
Looks like they're not enums: https://github.com/microsoft/fluentui-blazor/tree/main/src/Assets/FluentUI.Icons/Icons

This method might help, though it doesn't look cheap so results should be cached:

https://github.com/microsoft/fluentui-blazor/blob/main/src/Assets/FluentUI.Icons/Icons.cs#L20

You pass it an instance of [`IconInfo`](https://github.com/microsoft/fluentui-blazor/blob/dev/src/Core/Components/Icons/IconInfo.cs) which takes name, size and variant (filled/regular).

davidfowl replied:
I decided to stick with fixed names.

drewnoakes commented on src/Aspire.Hosting.Azure/Provisioning/Provisioners/BicepProvisioner.cs:
```diff
@@ -152,6 +154,13 @@ await foreach (var kv in keyVaults.GetAllAsync(cancellationToken: cancellationTo
+                {
+                    State = "Provisioning Keyvault",
+                    StateStyle = "info"
+
+                }).ConfigureAwait(false);
```
To make consistent with the code below:

```suggestion
                })
                .ConfigureAwait(false);
```

timheuer commented on src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:
```diff
@@ -104,11 +98,21 @@ async Task AfterProvisionAsync(IAzureResource resource)
-                await SetStateAsync(resource, "FailedToStart").ConfigureAwait(false);
+                await UpdateStateAsync(resource, s => s with
+                {
+                    State = "Failed to Provision",
```
later question: these likely need to be localized?

davidfowl replied:
Yep.

davidfowl committed:
Make style updates a single field in the app model
- This avoid reusing old state styles by mistake when changing text
- Added KnownResourceStyles

davidfowl committed:
Fixed proto comments

davidfowl commented:
/azp run

davidfowl commented:
/backport to release/8.0-preview5
